Tests: Use self.chain instead of 'regtest' in all current tests

jtimon opened the pull request:
Simply avoiding the hardcoded string in more places for consistency.
It can also allow for more easily reusing tests for other chains other than regtest.

Separated from #8994 .
Continues #16509 .

It is still not complete (ie to be complete, we need the -chain parameter in #16680 and make whether acceptnonstdtxs is allowed for that chain or not customizable for regtest [or for custom chains like in #8994 ] ). But while being incomplete like #16509 , it's quite simple to review and another step forward IMO.

jtimon committed:
Tests: Use self.chain instead of 'regtest' in almost all current tests

Sjors commented:
Concept ACK

jtimon commented:
Now also in newly included wallet_reorgsrestore.py

laanwj commented:
I don't understand the rationale behind this. How much chance do these tests have of working with a non-regtest chain?

instagibbs commented:
@laanwj with a custom testchain, most tests should work out of the box. Only ones that rely on specific genesis block information will break(there are a couple).

By itself it isn't greatly motivated, but it's also not a burden to maintain, so for me concept ACK.

laanwj commented:
Many of the functional tests depend on specific parameters of the regtest chain to be able to test things: reward schedule, mining difficulty, even address format.

Anyhow, conecpt ACK on this change, I was just wondering.

jtimon commented:
> I don't understand the rationale behind this. How much chance do these tests have of working with a non-regtest chain?

Well, any custom chain created with #8994 should pass all the tests, like, for example, does -chain=regtest2 (in that PR).
Other chains like signets I guess don't benefit much from this change.
Projects based on bitcoin core that periodically rebase like elements should benefit too (although, yes, they will need further changes, this won't be enough for them).

I guess separating it is mostly about:

1) Making #8994 smaller if this get merged first so that it's easier to review and maintain

2) hopefully new tests will copy the self.chain from existing tests instead of copying the "regtest". That would mean less work when rebasing #8994 

In summary, it's basically about #8994 , so if it doesn't make sense separated, I understand too.

laanwj commented:
Well I think in any case it's better to have the chain configurable instead of hardcoding a magic string `"regtest"`.

ACK 72d822fc2c46d55d959cc8d2f50145d8607e7126

MarcoFalke commented:
+-0 `self.chain` and `regtest` are the same for these test and I don't think it will ever change

jtimon commented:
@MarcoFalke does that imply a nack to #8994 ?

jtimon commented:
Also, wouldn't it make sense alone, if anything, for consistency with https://github.com/bitcoin/bitcoin/pull/16509 ?

MarcoFalke commented:
> does that imply a nack to #8994 ?

Yeah, I am -0 on changing to `regtest2`.

> Also, wouldn't it make sense alone, if anything, for consistency with #16509 ?

Yeah, +0 on this.

So overall +-0

Sjors commented:
ACK 72d822f. Tests still pass on macOS 10.14.6.

MarcoFalke commented:
Also, this is not a complete fix, since it is impossible to protect against re-introduction of `regtest` in new code. E.g. https://github.com/bitcoin/bitcoin/blob/eeecdfa27a14df83c441cb4b04c0ae78a2ccbd2c/test/functional/feature_loadblock.py#L39

MarcoFalke commented:
So, I'll change to overall -0.1

laanwj commented:
> Also, this is not a complete fix, since it is impossible to protect against re-introduction of regtest in new code. E.g. 

Strictly you could: randomize the test chain name every time.
Not sure it's worth it though …

> Yeah, I am -0 on changing to regtest2.

I don't think we should change the default name of the regtest network either. The only chain we'd potentially rename would be testnet, if the rules change, but an instance of regtest is meant to be short-lived.

jtimon commented:
> Also, this is not a complete fix,...

Rebased on top of master, but I don't see that test in master. It seems that commit doesn't belong in master unless I'm missing something.

I added https://github.com/jtimon/bitcoin/commit/e889b4ba1fc20fe49cce2f23f0594fd8bf19993d#diff-6651d2553d181295a8570f18069e6b2bR28 for completion when doing the grep.
In any case, even if it's not complete (and it's not complete because it lacks something like https://github.com/jtimon/bitcoin/commit/0943ce09151003daf4fc353687e6c084416d0b2a#diff-2a344479dbbded9df5d6a4abde2cd48cL44), it makes #16509 more complete.

> I don't think we should change the default name of the regtest network either.

I created https://github.com/bitcoin/bitcoin/pull/17032 and https://github.com/bitcoin/bitcoin/pull/17037 separated from #8994 that unlike #8994 don't change the default self.chain to regtest2. Neither of them (https://github.com/bitcoin/bitcoin/pull/17032 and https://github.com/bitcoin/bitcoin/pull/17037) needs this, but I still think it's worth it for consistency and to complete  #16509.

And to reiterate @laanwj 's words: 

> Many of the functional tests depend on specific parameters of the regtest chain to be able to test things: reward schedule, mining difficulty, even address format.

But if those reasons aren't enough and there's little chances of changing the default "regtest" to "regtest2" in the tests, then #8994 is not a justification for this either (by the way,  you can go nack it for that reason, I guess). In that case, I guess we should close this.

Or perhaps we can wait for https://github.com/bitcoin/bitcoin/pull/17032 or https://github.com/bitcoin/bitcoin/pull/17037 to be merged first for this to be able to be complete.

To be honest, what I care the most about personally now is #17037, not #8994 anymore.

MarcoFalke commented:
> In that case, I guess we should close this.

Ok, closing then

jtimon commented:
So consistency and completion are not good enough reasons to do things and neither is avoiding hardcoding strings?
Ok, noted.
Not sure why #16509 was merged then since it doesn't even do what it claims, but ok.

MarcoFalke commented:
> So consistency and completion are not good enough reasons to do things and neither is avoiding hardcoding strings?

As mentioned previously, this is not complete nor consistent. If you rebase on master, new `regtest` strings have been introduced. And there is no sane way to prevent that happen in the future, even if this pull was merged.

Also, we have a lot more stuff hardcoded to regtest in the functional tests, e.g. the bech32 hrp in `ADDRESS_BCRT1_UNSPENDABLE`...

jtimon commented:
As answered before, is (or was) only not complete for one case whuch is
contemplated in other prs.
The example you poibted out wasn't ecen part of master.

Of course, the more we delay this the more new things there will be to
correct, for people will continue to copy "regtest" instead of self.chain
from other tests.

I will grep again later, but I really hope this time you're not talking
about other branches again.
Feel free to poibt out which cases make this incomplete according to you.





On Fri, Oct 25, 2019, 19:15 MarcoFalke <notifications@github.com> wrote:

> So consistency and completion are not good enough reasons to do things and
> neither is avoiding hardcoding strings?
>
> As mentioned previously, this is not complete nor consistent. If you
> rebase on master, new regtest strings have been introduced. And there is
> no sane way to prevent that happen in the future, even if this pull was
> merged.
>
> Also, we have a lot more stuff hardcoded to regtest in the functional
> tests, e.g. the bech32 hrp in ADDRESS_BCRT1_UNSPENDABLE...
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/16681?email_source=notifications&email_token=AAHWGSV6LZ7PC7YALNVC6ULQQMSTVA5CNFSM4IOVRCR2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECI7JZI#issuecomment-546436325>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHWGSSCEHRC5ZGZL64DZB3QQMSTVANCNFSM4IOVRCRQ>
> .
>

jtimon commented:
Rebased, adapted new test feature_loadblock.py

Sjors commented:
re-ACK 1abcecc. I think it's an improvement even if incomplete and if some PR's might accidentally bring "regtest" back. Subsequent improvements hopefully don't have to touch 16 files.

I'd like to break the circular arguments between this and related pull requests. Those are complicated enough to reason about without "but I don't like this name change in the other PR"; using `self.chain` instead of `"regtest"` gives us more optionality.

ryanofsky reviewed (approved):
Code review ACK 1abcecc40c518a98b7d17880657ec0247abdf125.

I think one reason a change like this can get stuck in the review process is not stating a motivation or list of practical advantages in the PR description. When a description begins with "Separated from _. continues _. It is incomplete...", it suggests a rabbit hole few people will want to go down.

On the other hand, even though our review process is intentionally conservative, I think if a change is safe and has had sufficient code review that we shouldn't hold neutral or "+0" feedback against it. If as a reviewer you think drawbacks and risks of a change outweigh its benefits, you should actually state that, so we can avoid uncertainty and open PRs sitting in limbo.

ryanofsky reviewed (approved):
Code review ACK 1abcecc40c518a98b7d17880657ec0247abdf125

jtimon commented:
I guess if I rebase now there will be more instances, but then I would lose the 2 ACKs.
What should I do?

elichai commented:
Code review ACK 1abcecc40c518a98b7d17880657ec0247abdf125
I do wonder how many of these tests will fail on testnet (even a custom one with difficulty 1).
Anyhow reducing hard coded strings is good IMHO, and I think even though it's incomplete it will help prevent future hard coded strings just by the fact that a lot of people copy-paste that code.

ryanofsky commented:
> I guess if I rebase now there will be more instances, but then I would lose the 2 ACKs.
> What should I do?

I'd suggest rebasing, the changes here are straightforward and I'm happy to reack.

MarcoFalke commented:
This has two ACKs, merging now.

fanquake commented:
Opened #18068 to track a followup PR.
